src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java

```diff
@@ -250,99 +263,48 @@ public void addToRowFilter(RowFilter.Builder filter,
             filter.add(columnDef, Operator.IN, buffer);
         }
 
-        @Override
-        protected final boolean isSupportedBy(Index index)
-        {
-            return index.supportsExpression(columnDef, Operator.IN);
-        }
-
-        protected abstract List<ByteBuffer> getValues(QueryOptions options);
-    }
-
-    public static class InRestrictionWithValues extends INRestriction
-    {
-        protected final List<Term> values;
-
-        public InRestrictionWithValues(ColumnMetadata columnDef, List<Term> values)
-        {
-            super(columnDef);
-            this.values = values;
-        }
-
-        @Override
-        MultiColumnRestriction toMultiColumnRestriction()
-        {
-            return new MultiColumnRestriction.InRestrictionWithValues(Collections.singletonList(columnDef), values);
-        }
-
         @Override
         public void addFunctionsTo(List<Function> functions)
         {
-            Terms.addFunctions(values, functions);
+            terms.addFunctionsTo(functions);
         }
 
         @Override
-        protected List<ByteBuffer> getValues(QueryOptions options)
+        public final boolean isSupportedBy(Index index)
         {
-            List<ByteBuffer> buffers = new ArrayList<>(values.size());
-            for (Term value : values)
-                buffers.add(value.bindAndGet(options));
-            return buffers;
+            return index.supportsExpression(columnDef, Operator.IN);
         }
 
         @Override
         public String toString()
         {
-            return String.format("IN(%s)", values);
+            return String.format("IN(%s)", terms);
         }
     }
 
-    public static class InRestrictionWithMarker extends INRestriction
+    public static class SliceRestriction extends SingleColumnRestriction
     {
-        protected final AbstractMarker marker;
+        private final TermSlice slice;
+        private final List<MarkerOrTerms> skippedValues; // values passed in NOT IN
```

Comment:
Is this supposed to work for `!=`? I tested this out with `!=` and it uses this class, but doesn't seem to work correctly. When I update the `isSupportedBy` method, I don't seem to get the right results, so perhaps it is intentional to leave out that support right now.

Here is my test:

```java
    @Test
    public void notEqualsShouldReturnRowsWithNullValues() throws Throwable
    {
        createTable("CREATE TABLE %s(id int PRIMARY KEY, num int)");
        createIndex("CREATE CUSTOM INDEX ON %s(num) USING 'StorageAttachedIndex'");
        waitForIndexQueryable();
        execute("INSERT INTO %s(id, num) values (1, 1)");
        execute("INSERT INTO %s(id) values (3)");
        flush();
        execute("INSERT INTO %s(id) values (2)");

        beforeAndAfterFlush(() -> {
            assertRows(execute("SELECT id FROM %s WHERE num != 1"), row(2), row(3));
        });
    }
```

Reply:
Why are we supposed to return nulls in this case?

Reply:
It wasn't working at all even after fixing `isSupportedBy` method (I was hitting NPE due to partitionKey component of a `PrimaryKey` being null). Fixed.